stock-portfolio-monitor/src/market/MarketState.ts

```diff
@@ -0,0 +1,269 @@
+
+    return Promise.all(this.portfolioState.uniqueTickers.map(async ticker => {
+      const allCandles = await loadCandles(ticker, initialDateFrom.atTime(LocalTime.MIDNIGHT).toInstant(ZoneOffset.UTC),
+        todayInstant, this.period)
+      const firstIndex = allCandles.findIndex(_ => _.instant.epochSecond() >= dateFromInstant.epochSecond())
+      let candles
+      if (firstIndex < 2) {
+        candles = allCandles
+      } else {
+        candles = allCandles.slice(firstIndex - 1)
+      }
+      return {
+        ticker: ticker,
+        candles: candles
+      }
+    }))
+  }
+
+  private calculateDateFrom(dateTo: LocalDate): LocalDate {
+    let dateFrom
+    switch (this.period) {
+      case ViewPeriod.Week:
+        dateFrom = dateTo.minusDays(7)
+        break
+      case ViewPeriod.Month:
+        dateFrom = dateTo.minusMonths(1)
+        break
+      case ViewPeriod.Year:
+        dateFrom = dateTo.minusYears(1).withDayOfMonth(1)
+        break
+    }
+    return dateFrom
+  }
+
+  private updateSelectedPeriodInfo(stockProfits: StockProfit[], candleGroups: TickerCandleGroup[], profitPoints: ProfitPoint[]) {
+    let info = this.periodInfo.find(_ => _.period === this.period)
+    let infoCreated = false
+    if (!info) {
+      info = { period: this.period, stockProfits: [], candleGroups: [], profitPoints: [] }
+      infoCreated = true
+    }
+    info.stockProfits = stockProfits
+    info.candleGroups = candleGroups
+    info.profitPoints = profitPoints
+    if (infoCreated) {
+      this.periodInfo.push(info)
+    }
+  }
+
+  private calculateStockProfits(candleGroups: TickerCandleGroup[]): StockProfit[] {
+    return this.portfolioState.uniqueTickers.map(ticker => {
+      const firstCandle = this.findFirstCandle(ticker, candleGroups)
+      let value = 0
+      let percent = 0
+      let previousInvestedValue = 0
+      if (firstCandle) {
+        const currentPrice = this.currentPrices.find(_ => _.ticker === ticker)?.price
+        if (currentPrice !== undefined) {
+          const firstDate = firstCandle.instant.atZone(ZoneOffset.UTC).toLocalDate()
+          const items = this.portfolioState.items.filter(_ => _.ticker === ticker)
+          items.forEach(item => {
+            const previousPrice = item.date.isBefore(firstDate) ? firstCandle.close : item.price
+            previousInvestedValue += previousPrice * item.count
+            value += (currentPrice - previousPrice) * item.count
+          })
+          if (previousInvestedValue > 0) {
+            percent = value / previousInvestedValue * 100
+          }
+        }
+      }
+      return {
+        ticker: ticker,
+        value: value,
+        percent: percent,
+        previousInvestedValue: previousInvestedValue
+      }
+    })
+  }
+
+  private findFirstCandle(ticker: Ticker, candleGroups: TickerCandleGroup[]): Candle | undefined {
+    return candleGroups.find(_ => _.ticker === ticker)?.candles?.[0]
```

Comment:
`_ => _.ticker === ticker`: `_` is confusing all the time 😅 Don't use it anywhere, please.